content/blog/what-are-uuids/index.md

```diff
@@ -0,0 +1,228 @@
+
+- **Easy to guess the sequence**: Likewise, let's say you want to use a UUID to generate a reset password link to the user:
+
+  ```
+  https://example.com/reset/ef5dc6d0-93ca-11ed-acdd-6fb6f98f3ada
+  ```
+
+  An attacker can decode the Mac address of the server (in this case, `6f:b6:f9:8f:3a:da`) and use it to brute force the rest of the URL, as demonstrated [in this article by Versprite](https://versprite.com/blog/universally-unique-identifiers/).
+
+## UUIDv2s are _weird_ {#UUIDv2}
+
+> I know you mentioned it's a long story with UUIDv2 when outlining the different versions of UUID... Surely it can't be that bad, can it?
+
+No, really, it's a ***long*** story. It's complex enough that I'm writing a second blog post explaining what UUIDv2s are, why they're not widely used, and how they came to be.
+
+In the meantime, the short version of UUIDv2 is:
+
+----
+
+UUIDv2 matches the specification for UUIDv1 but replaces some of the timing and clock information with identification numbers for operating system usage.
+
+They're rarely implemented into most UUID libraries and are used even more rarely. This is because of 2 reasons:
+
+1) They're not documented as part of the newest UUID specification
+2) Significant pitfalls in UUIDv2's generation schema that **regularly** leads to collision with other generated UUIDv2s
+
+As a result, we won't talk about UUIDv2s anymore today. Stay tuned for an upcoming article outlining them more.
+
+## Namespace Your IDs with UUIDv3 and UUIDv5 {#UUIDv3and5}
+
+Let's say you're creating a database of URLs tracked on your website. You want to use the URL as [the primary key in your database](https://en.wikipedia.org/wiki/Primary_key) but don't want to use the full URL, as it will bloat your key's storage requirements.
+
+On the other side of the coin: You can't use any random data to store as your URL's primary key since it could introduce duplicate URL entries into your database table. 
+
+Is there a good way to generate a unique ID based on a string that will output the same ID if the inputs are the same?
+
+Intro: UUIDv3 and UUIDv5.
+
+Both of these UUID versions take the following information:
+
+- A pre-defined UUID "namespace"
+- Any string for a "name"
+
+> The UUID specification establishes 4 pre-defined namespaces for common use cases. The pre-defined namespaces are:
+>
+> - [DNS](https://en.wikipedia.org/wiki/Domain_Name_System): `6ba7b810-9dad-11d1-80b4-00c04fd430c8`
+> - [URL](https://en.wikipedia.org/wiki/URL): `6ba7b811-9dad-11d1-80b4-00c04fd430c8`
+> - [OID](https://en.wikipedia.org/wiki/Object_identifier): `6ba7b812-9dad-11d1-80b4-00c04fd430c8`
+> - [X.500 DN](https://en.wikipedia.org/wiki/X.500): `6ba7b814-9dad-11d1-80b4-00c04fd430c8`
+
+These UUID versions output a UUID that contains a hash of the namespace and name concatenated together.
+
+You can think of the generation algorithm for both of these UUID versions as the following:
+
+```javascript
+UUID = hash(NAMESPACE + NAME)
+```
+
```

Comment:
**Additional explanation:**
The reason a namespace is needed (instead of just hashing the name alone) is that even if the name is unique within a certain set of items (or "namespace" of items) it's likely that two different items from two different namespaces share the same name.
So to avoid duplicate UUIDs across ALL namespaces, we add the namespace UUID and hash the combination.
_And that allows us to generate the same UUID with the same input each time, which means we don't even have to store the value, we can generate them on the go._